File: sol/solution/0001-0100/0011. Container With Most Water/Solution.cpp
class Solution
{
public:
    int maxArea(vector<int> &height)
    {
        int left = 0;
        int right = height.size() - 1;
        int max_water = 0;
        while (left < right)
        {
            max_water = max(min(height[left], height[right]) * (right - left), max_water);
            if (height[left] <= height[right])
            {
                left++;
            }
            else
            {
                right--;
            }
        }
        return max_water;
    }
};

File: sol/solution/0001-0100/0011. Container With Most Water/Solution.erl
-spec max_area(Height :: [integer()]) -> integer().
max_area(Height) ->
    Tuple = list_to_tuple(Height),
    max_area(Tuple, 0, tuple_size(Tuple) - 1, 0).
max_area(Tuple, Left, Right, MaxWater) when Left < Right ->
    Width = Right - Left,
    LeftHeight = element(Left + 1, Tuple),
    RightHeight = element(Right + 1, Tuple),
    Area = min(LeftHeight, RightHeight) * Width,
    NewMaxWater = max(MaxWater, Area),
    if
        LeftHeight < RightHeight ->
            max_area(Tuple, Left + 1, Right, NewMaxWater);
        true ->
            max_area(Tuple, Left, Right - 1, NewMaxWater)
    end;
max_area(_Tuple, _Left, _Right, MaxWater) ->
    MaxWater.